### PR TITLE
Spark/jdbc relations

### PR DESCRIPTION
**Pull request by tnazarew**
### Problem

As JDBC source is external, Spark doesn't have info about any columns that are not in DF. That wouldn't be an issue if only tables could be loaded but we can load with queries. Said query can have join operations, aggregations, operations on columns and so on.

Closes: #1048 

### Solution

Using OL sql parser to get information from queries, add it to info we currently get from Spark plan. That requires implementing some mapping between spark expressions and jdbc expression.

> **Note:** All schema changes require discussion. Please [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) for context.

- [ ] Your change modifies the [core](https://github.com/OpenLineage/OpenLineage/blob/main/spec/OpenLineage.json) OpenLineage model
- [ ] Your change modifies one or more OpenLineage [facets](https://github.com/OpenLineage/OpenLineage/tree/main/spec/facets)

If you're contributing a new integration, please specify the scope of the integration and how/where it has been tested (e.g., Apache Spark integration supports `S3` and `GCS` filesystem operations, tested with AWS EMR).

### Checklist

- [x] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [x] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [x] Your changes are accompanied by tests (_if relevant_)
- [x] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [x] Your comment includes a one-liner for the changelog about the specific purpose of the change (_if necessary_)
- [ ] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)
- [x] You've added a [header](https://github.com/OpenLineage/OpenLineage/tree/main/.github/header_templates.md) to source files (_if relevant_)

----
SPDX-License-Identifier: Apache-2.0\
Copyright 2018-2023 contributors to the OpenLineage project

### CODE REVIEW

File: integration/spark/app/build.gradle [hunk 1]
```diff
@@ -202,9 +202,9 @@ task copyDependencies(type: Copy) {
 task integrationTest(type: Test) {
     // integrationTest depends on shadowJar which dependsOn test
     // exclude all the tests and add manually tagged with integration-test
-    test {
-        exclude "**"
-    }
+//    test {
```
Comment (tnazarew) [inline integration/spark/app/build.gradle:205]: someone might check this, but this line also disables tests in `test` step for me

File: integration/spark/spark3/src/main/java/io/openlineage/spark3/agent/lifecycle/plan/column/InputFieldsCollector.java [hunk 2]
```diff
@@ -60,13 +68,63 @@ static void collect(
 
   private static void discoverInputsFromNode(
       OpenLineageContext context, LogicalPlan node, ColumnLevelLineageBuilder builder) {
-    extractDatasetIdentifier(context, node).stream()
+    List<DatasetIdentifier> datasetIdentifiers = extractDatasetIdentifier(context, node);
```
Comment (pawel-big-lebowski) [inline integration/spark/spark3/src/main/java/io/openlineage/spark3/agent/lifecycle/plan/column/InputFieldsCollector.java:71]: I would be really in favor of `JdbcColumnLineageCollector` with methods `extractExternalInputs` and `extractExpressionsFromJDBC` with some unit tests attached it.

**Review by pawel-big-lebowski — COMMENTED** [review]
I am really happy to see that working 👍 🥇 💯 

It's a substantial amount of code that deserves more than just a single integration test added. 
Unit tests are really useful to detect problems quickly when refactoring. Additionally, the code contains multiple conditions which may make sense to be checked. 

How about storing data to jdbc? Would it require any additional work to have column lineage supported?

File: integration/spark/app/integrations/container/columnLineageJDBCStart.json [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            },
+            "value": {
+              "inputFields": [
+                {
+                  "namespace": "file",
+                  "name": "/tmp/iceberg/default.v2_source_2",
+                  "field": "v2"
+                },
+                {
+                  "namespace": "jdbc",
```
[the remark this answers is not in the document]
Comment (tnazarew) [reply integration/spark/app/integrations/container/columnLineageJDBCStart.json:128]: shoud it be? if so then its no problem to add it

File: integration/spark/spark3/src/main/java/io/openlineage/spark3/agent/lifecycle/plan/column/ColumnLevelLineageBuilder.java [hunk 4]
```diff
@@ -205,4 +207,12 @@ private List<ExprId> findDependentInputs(ExprId outputExprId) {
 
     return dependentInputs;
   }
+
+  public void addExternalMapping(ColumnMeta meta, ExprId exprid) {
+    externalExpressionMappings.putIfAbsent(meta, exprid);
+  }
+
+  public ExprId getMappig(ColumnMeta columnMeta) {
```
[the remark this answers is not in the document]
Comment (tnazarew) [reply integration/spark/spark3/src/main/java/io/openlineage/spark3/agent/lifecycle/plan/column/ColumnLevelLineageBuilder.java:215]: maybe, but which one exactly?

File: integration/spark/app/build.gradle (hunk 1 above)
Comment (pawel-big-lebowski) [inline integration/spark/app/build.gradle:205]: ?

File: integration/spark/app/integrations/container/columnLineageJDBCStart.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (pawel-big-lebowski) [reply integration/spark/app/integrations/container/columnLineageJDBCStart.json:128]: https://github.com/OpenLineage/OpenLineage/blob/main/spec/Naming.md#data-warehousesdata-bases

[the remark this answers is not in the document]
Comment (tnazarew) [reply integration/spark/app/integrations/container/columnLineageJDBCStart.json:128]: will be done

File: integration/spark/spark3/src/main/java/io/openlineage/spark3/agent/lifecycle/plan/column/ColumnLevelLineageBuilder.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (pawel-big-lebowski) [reply integration/spark/spark3/src/main/java/io/openlineage/spark3/agent/lifecycle/plan/column/ColumnLevelLineageBuilder.java:215]: mappiNg?

File: integration/spark/app/build.gradle (hunk 1 above)
Comment (tnazarew) [reply integration/spark/app/build.gradle:205]: it was done so the tests were not omitted in `app`, was supposed to be rolled back

File: integration/spark/spark3/src/main/java/io/openlineage/spark3/agent/lifecycle/plan/column/ColumnLevelLineageBuilder.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (tnazarew) [reply integration/spark/spark3/src/main/java/io/openlineage/spark3/agent/lifecycle/plan/column/ColumnLevelLineageBuilder.java:215]: oh yeah

**Review by mobuchowski — APPROVED** [review]

**Pull request merged**